crates/typst/src/eval/call.rs

```diff
@@ -293,47 +281,92 @@ pub(crate) fn call_closure(
 /// associated functions would make an addition of a new associated function a breaking
 /// change and prioritizing fields would break associated functions for certain
 /// dictionaries).
-fn eval_callee_args(
-    target: Value,
-    target_span: Span,
-    field: Ident,
-    mut args: Args,
-) -> SourceResult<(Value, Args)> {
-    if let Some(callee) = target.ty().scope().get(&field) {
-        args.insert(0, target_span, target);
-        Ok((callee.clone(), args))
+fn eval_method_call_and_args(
+    target_expr: ast::Expr,
+    method: Ident,
+    args: ast::Args,
+    span: Span,
+    vm: &mut Vm,
+) -> SourceResult<EvaluatedMethod> {
+    // Evaluate the method's target and overall arguments.
+    let (target, mut args) = if is_mutating_method(&method) {
+        // If this looks like a mutable method then we evaluate the arguments before the
+        // target because calling `target_expr.access(vm)` mutably borrows the vm, meaning
+        // we won't be able to evaluate args after due to the borrow checker.
+        let args = args.eval(vm)?.spanned(span);
+        // However, this difference from the normal call order is not observable because
+        // expressions like, i.e. `(1, arr.len(), 2, 3).push(arr.pop())`, evaluate the
+        // target to a temporary which we disallow mutation on (returning an error).
+        // Theoretically this could be observed if a method matching `is_mutating_method`
+        // was added to some type in the future and we didn't update this function.
+        match target_expr.access(vm)? {
+            // Only arrays and dictionaries have mutable methods.
+            target @ (Value::Array(_) | Value::Dict(_)) => {
+                let value = call_method_mut(target, &method, args, span);
+                let point = || Tracepoint::Call(Some(method.get().clone()));
+                return Ok(EvaluatedMethod::Special(value.trace(vm.world(), point, span)?));
+            }
+            target => (target.clone(), args),
+        }
+    } else {
+        // This is consistent with evaluating the callee first and args second as long as
+        // finding the method on the target is not affected by evaluating the args, which
+        // I believe is true
```

Comment:
I think this comment can be removed, as the code statements below explain themselves already.

If we are not sure whether `target_expr.eval(vm)` affects the result of `args.eval(vm)`, I think we could comment the typst PR (not here) and I believe the maintainers will help us.

Reply:
Not sure, I feel like I want to point out why the order is the way it is, maybe just shorten the comment to something like this?
> Function call order: we evaluate the method target before the arguments, but look up the method after.

Reply:
Or leave it here for maintainers to judge.

Reply:
On second thought, I realize this is now commented above in the function's doc comment, so I'm just going to remove it.